### PR TITLE
Deprecate `Grab Existential Variables` and `Existential` commands

### PR DESCRIPTION
**Pull request by maximedenes**
I thought this was done a long time ago, but it seems that these commands were never actually deprecated.

I'll update the changelog and documentation if the PR is approved.

Related: https://github.com/coq/coq/issues/9545, https://github.com/coq/coq/issues/9171

**Kind:** documentation / bug fix / feature / performance / infrastructure.

Fixes / closes #????

- [ ] Added / updated test-suite

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by Zimmi48** [comment]
FTR, the reason why this was not done before (for instance, I also wanted to deprecate them) is because there remains some bugs that make some existential variables escape the shell. I'm OK to deprecate right now though, but the warning should probably mention something about reporting bugs if `Unshelve` is not sufficient.

**Comment by maximedenes** [comment]
> should probably mention something about reporting bugs if `Unshelve` is not sufficient.

Not clear to me what is specific to this kind of bugs (we already encourage to report all bugs), but I'd be happy to add any information to the warning.

**Comment by maximedenes** [comment]
> there remains some bugs that make some existential variables escape the shell

do you have pointers to such issues btw?

**Comment by Zimmi48** [comment]
What I mean is that we should clearly communicate to the users that we don't plan to remove a command that's useful to them before having fixed all the associated bugs that they report (similar to what happened with the `omega` deprecation / `lia` replacement).

**Comment by Zimmi48** [comment]
The other thing that is specific to this kind of bugs is that it may not be obvious to users that these are actually bugs.

**Comment by Zimmi48** [comment]
Here's a previous discussion of the related issues: #9170

**Comment by Zimmi48** [comment]
Maybe in fact these bugs no longer exist, but we cannot know if users don't report them.

**Comment by maximedenes** [comment]
> The other thing that is specific to this kind of bugs is that it may not be obvious to users that these are actually bugs.

Really? `Unshelve` not unshelving some evars seems pretty straightforward, no? Or maybe I misunderstood the king of bugs we are talking about.

> Here's a previous discussion of the related issues: #9170

I already had a look at it, but didn't see any issue about evars captured by `Grab Existential Variables` but not by `Unshelve`. I guess we are talking about different things. What kind of bugs did you have in mind?

**Comment by maximedenes** [comment]
> Maybe in fact these bugs no longer exist, but we cannot know if users don't report them.

But that's exactly the point of deprecating (as opposed to removing), no? I'm completely confused now about what happened.

**Comment by Zimmi48** [comment]
Note that I don't have a specific bug in mind, but I am just repeating the general folklore related to these commands. E.g., in this PR, @ppedrot said:
>Just to say it out loud: in an ideal world these commands should be removed, but they're fairly handy when it comes to working around Coq bugs.

>But that's exactly the point of deprecating (as opposed to removing), no? I'm completely confused now about what happened.

Let me be clear: I support this deprecation.

> > The other thing that is specific to this kind of bugs is that it may not be obvious to users that these are actually bugs.
> 
> Really? `Unshelve` not unshelving some evars seems pretty straightforward, no? Or maybe I misunderstood the king of bugs we are talking about.

> > Here's a previous discussion of the related issues: #9170
> 
> I already had a look at it, but didn't see any issue about evars captured by `Grab Existential Variables` but not by `Unshelve`. I guess we are talking about different things. What kind of bugs did you have in mind?

Maybe it is wrong, but I have a vague memory of @JasonGross asking whether this kind of behavior (`Grab Existential Variables` giving more goals than `Unshelve`) should always be considered a bug or not. If Jason had doubts about that, I bet that most users do as well. The thing is that without clear specifications, it's a bit hard for users to know for sure when a weird behavior is a bug.

**Comment by maximedenes** [comment]
I see, thanks for the clarification.

**Comment by JasonGross** [comment]
I also have a vague memory of asking, though I'm not sure where.  I reference the answer I got at https://github.com/coq/coq/issues/9171#issuecomment-445512132 though.

Also, here is another related bug I've not seen mentioned: https://github.com/coq/coq/issues/11363

**Comment by maximedenes** [comment]
@coq/engine-maintainers do you approve this deprecation?

**Comment by ppedrot** [comment]
I personally believe it's still a bit early, but I won't oppose either.

**Comment by maximedenes** [comment]
> I personally believe it's still a bit early

What step should be completed first, then?

**Comment by ppedrot** [comment]
> What steps should be completed first, then?

Fixing all the stray evar bugs and settling for a clear shelving semantics? I agree we should signal that these commands should not be used in a normal workflow, but deprecation is a promise to remove them soonish, while it's unlikely for all the shelving bugs to be solved that fast.

**Comment by maximedenes** [comment]
> Fixing all the stray evar bugs

AFAICT there's only one known bug, which is #11363 as mentioned by Jason. I'm ok with fixing it before merging this PR.

> settling for a clear shelving semantics

How does this relate to the two commands discussed here? The shelving tactics/commands are already available and used.

What kind of clarification would you like to see on the semantics?

**Comment by ppedrot** [comment]
> What kind of clarification would you like to see on the semantics?

Nah, fine, go ahead, I am OK with this PR.

**Comment by gares** [comment]
This PR needs a changelog entry at least. Otherwise @ppedrot can you approve/assign?

**Comment by Zimmi48** [comment]
Needs also a `.. deprecated:: 8.13` in the documentation of these two commands.

**Comment by ppedrot** [comment]
Not sure whether it's legal as an assignee but I can write a quick changelog?

**Comment by Zimmi48** [comment]
I think it's fine. I can review what you write. Please also consider the deprecation annotations in the doc as I suggested above.

**Comment by ppedrot** [comment]
Ready.

File: doc/changelog/07-commands-and-options/12516-deprecate-grab-existentials.rst
Comment (Zimmi48) [inline doc/changelog/07-commands-and-options/12516-deprecate-grab-existentials.rst:2]:
```suggestion
  :cmd:`Grab Existential Variables` and :cmd:`Existential commands`
```

File: doc/changelog/07-commands-and-options/12516-deprecate-grab-existentials.rst
Comment (Zimmi48) [inline doc/changelog/07-commands-and-options/12516-deprecate-grab-existentials.rst:4]:
```suggestion
  by Maxime Dénès).
```

File: doc/sphinx/proofs/writing-proofs/proof-mode.rst
Comment (Zimmi48) [inline doc/sphinx/proofs/writing-proofs/proof-mode.rst:299]:
```suggestion
   .. deprecated:: 8.13

      Use :cmd:`Unshelve` instead.
```

File: doc/changelog/07-commands-and-options/12516-deprecate-grab-existentials.rst
Comment (ppedrot) [reply doc/changelog/07-commands-and-options/12516-deprecate-grab-existentials.rst:4]: Aha, the script is trying to attribute me work that is not mine!

**Review by Zimmi48 — APPROVED** [review]

**Comment by ppedrot** [comment]
Waiting for the refman check to run and will merge right after.

**Review by ppedrot — APPROVED** [review]
engine-maintainers Official Seal of Approval.

**Comment by maximedenes** [comment]
> Not sure whether it's legal as an assignee but I can write a quick changelog?

@ppedrot Thanks a lot for your help, it is very welcome in these busy times :)

**Pull request merged**
